File: 02.py
#! /usr/bin/env python3

import re

input = [l for l in open('02.in').read().split('\n') if l]

valid1 = valid2 = 0
for i in input:
    range, letter, pwd = i.split()
    lo, hi = map(int, range.split('-'))
    letter = letter[0]
    if lo <= pwd.count(letter) <= hi:
        valid1 += 1
    if (pwd[lo-1] + pwd[hi-1]).count(letter) == 1:
        valid2 += 1
print(valid1, valid2)

File: 02.rs
mod aoc;

fn str_count(s: &str, letter: char) -> usize {
    s.chars().filter(|l| *l == letter).count()
}

fn main() {
    let input = aoc::lines_from_file("02.in");
    let mut valid1 = 0;
    let mut valid2 = 0;

    for i in input {
        let parts: Vec<&str> = i.split_whitespace().collect();
        let limits: Vec<usize> = parts[0].split('-').filter_map(|l| l.parse().ok()).collect();
        let lo = limits[0];
        let hi = limits[1];
        let letter = parts[1].chars().nth(0).unwrap();
        let pwd = parts[2];

        let c = str_count(pwd, letter);
        if lo <= c && c <= hi {
            valid1 += 1
        }

        let mut neighbors = String::from("");
        neighbors.push(pwd.chars().nth(lo - 1).unwrap());
        neighbors.push(pwd.chars().nth(hi - 1).unwrap());
        if str_count(&neighbors, letter) == 1 {
            valid2 += 1
        }
    }
    println!("{} {}", valid1, valid2)
}
